devmapper: Return device id in error message

rhvgoyal opened the pull request:
I often get complaints that container removal failed and users got following
error message.

"Driver devicemapper failed to remove root filesystem 18a69ba82aaf7a039ce7d44156215012d703001643079775190ac7dd6c6acf56:Device is Busy"

This error message talks about container id but does not give any info
about which particular device id is busy. Most likely device is mounted
in some other mount namespace and if one knows the device id, they
can try to do some debugging figuring which process and which mount
namespace is keeping the device busy and how did we reach that stage.

Without that information, it becomes almost impossible to debug the
problem.

So to improve the debuggability, when device removal fails, also return
device id in error message. Now new message looks as follows.

"Driver devicemapper failed to remove root filesystem 18a69ba82aaf7a039ce7d44156215012d703001643079775190ac7dd6c6acf56: Failed to remove device dbc15bdf9994a17c613d8ef9e924f3cffbf67f91e4f709295c901ad628377991:Device is Busy"

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
cc @thaJeztah

mrunalp commented on daemon/graphdriver/devmapper/driver.go:
```diff
@@ -151,7 +151,7 @@ func (d *Driver) Remove(id string) error {
 	// This assumes the device has been properly Get/Put:ed and thus is unmounted
 	if err := d.DeviceSet.DeleteDevice(id, false); err != nil {
-		return err
+		return fmt.Errorf("Failed to remove device %v:%v", id, err)
```
nit: Failed -> failed (go style convention)

rhvgoyal replied:
ok, will fix it.

rhvgoyal committed:
devmapper: Return device id in error message

I often get complains that container removal failed and users got following
error message.

"Driver devicemapper failed to remove root filesystem 18a69ba82aaf7a039ce7d44156215012d703001643079775190ac7dd6c6acf56:Device is Busy"

This error message talks about container id but does not give any info
about which particular device id is busy. Most likely device is mounted
in some other mount namespace and if one knows the device id, they
can try to do some debugging figuring which process and which mount
namespace is keeping the device busy and how did we reach that stage.

Without that information, it becomes almost impossible to debug the
problem.

So to improve the debuggability, when device removal fails, also return
device id in error message. Now new message looks as follows.

"Driver devicemapper failed to remove root filesystem 18a69ba82aaf7a039ce7d44156215012d703001643079775190ac7dd6c6acf56: Failed to remove device dbc15bdf9994a17c613d8ef9e924f3cffbf67f91e4f709295c901ad628377991:Device is Busy"

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@mrunalp fixed. PTAL.

mrunalp commented:
Looks good. Thanks!

tonistiigi commented:
LGTM

thaJeztah commented on daemon/graphdriver/devmapper/driver.go:
```diff
@@ -151,7 +151,7 @@ func (d *Driver) Remove(id string) error {
 	// This assumes the device has been properly Get/Put:ed and thus is unmounted
 	if err := d.DeviceSet.DeleteDevice(id, false); err != nil {
-		return err
+		return fmt.Errorf("failed to remove device %v:%v", id, err)
```
Perhaps import the `github.com/pkg/errors` package, and use;

```go
return errors.Wrapf(err, "failed to remove device %v", id)
```

rhvgoyal replied:
@thaJeztah Tried to have a quick look at errors.Wrapf() but it is not obvious to me what are we gaining by using errors.Wrapf() instead of fmt.Errorf(). Can you elaborate a bit.

thaJeztah replied:
We try to use it more by default. The `errors.Wrap` keeps the original error intact, which in cases can help to retrieve the underlying error, i.e., the _original_ error can be obtained through `errors.Cause(wrappedError)`. Maybe not used in this case, just adding more context to the error

thaJeztah commented:
Maybe additional information needs to be added to this document https://github.com/docker/docker.github.io/blob/master/engine/admin/troubleshooting_volume_errors.md to instruct people how to use this extra information

rhvgoyal commented:
@thaJeztah I see that about volumes and errors related to directory removal for shm. I don't think in both cases we will get this error as they don't go to graph drivers. 

So I don't think that this particular article is right place to put additional information. I would think that let us limit this PR to only add device id in error message and later another PR can come adding more documentation.

thaJeztah commented:
> and later another PR can come adding more documentation.

Yes, those docs are in another repository, so would have to be a separate PR anyway; I just recalled that addition to the documentation, and thought it was related to this

rhvgoyal commented:
@thaJeztah is errors.Wrapf() is a serious concern for you? If not, would you mind acking this PR and merging it.

tiborvass commented:
@rhvgoyal it is a serious concern indeed. We eventually want to eliminate all err-wrapping `fmt.Errorf()` instances in docker's code. It would have been a perfect opportunity to start the conversion since this PR only changes the error. But we can do that in a separate PR too.

LGTM

runcom commented:
Can this be included in 1.13.0 as well?

rhvgoyal commented:
Including it in 1.13 will be good.
